### PR TITLE
Upgrade probot from v5 to v9

### PR DESCRIPTION
**Pull request by jameesjohn**
## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

**Comment by ankita240796** [comment]
@DubeySandeep @bansalnitish pinging you for a review here. Thanks!

File: package.json [hunk 1]
```diff
@@ -14,8 +16,8 @@
     "google-auth-library": "^0.12.0",
     "googleapis": "^25.0.0",
     "newrelic": "^4.9.0",
-    "probot": "^5.0.0",
-    "probot-scheduler": "^1.2.0",
+    "probot": "^9.5.3",
+    "probot-scheduler": "^2.0.0-beta.1",
```
Comment (ankita240796) [inline package.json:20]: Why are we using the beta version?

File: package.json [hunk 2]
```diff
@@ -14,8 +16,8 @@
     "google-auth-library": "^0.12.0",
     "googleapis": "^25.0.0",
     "newrelic": "^4.9.0",
-    "probot": "^5.0.0",
-    "probot-scheduler": "^1.2.0",
+    "probot": "^9.5.3",
```
Comment (ankita240796) [inline package.json:19]: Why not 9.11.3, the latest version? (https://github.com/probot/probot/releases)

Comment (jameesjohn) [reply package.json]: Hmmm
How I did the update was I created a new probot app and used the version I found in the package.json generated by probot.

Updated.

File: package.json (hunk 1 above)
Comment (jameesjohn) [reply package.json:20]: I noticed that the previous version `1.2.0` had some issues after the upgrade, so I decided to upgrade it as well and currently, the beta version is the latest version.

**Comment by jameesjohn** [comment]
PTAL @ankita240796, I've responded to your comments. Thanks

File: package.json (hunk 1 above)
Comment (ankita240796) [reply package.json:20]: But isn't the beta version also a version for testing and can have issues - can we confirm this on their repo once - that is it safe to use the beta version?

Comment (jameesjohn) [reply package.json:20]: It's not stated that's it's unsafe, but there is an issue on the repository https://github.com/probot/scheduler/issues/35

Also, v1.2.0 of probot-scheduler doesn't work fine with probot v9

Comment (ankita240796) [reply package.json:20]: Ok, so here is what I would suggest: it is always better to confirm in the start. So, let us either create an issue or send a mail with our query regarding why is the latest release in beta version and is it ok to use it.

Comment (jameesjohn) [reply package.json:20]: Okay.
I've created an issue https://github.com/probot/scheduler/issues/41

Comment (ankita240796) [reply package.json:20]: Thanks!

File: index.js [hunk 3]
```diff
@@ -1,19 +1,22 @@
...
 const checkMergeConflictsModule = require('./lib/checkMergeConflicts');
 const checkPullRequestLabelsModule = require('./lib/checkPullRequestLabels');
 const whitelistedAccounts = (
   (process.env.WHITELISTED_ACCOUNTS || '').toLowerCase().split(','));
-var pullRequestAuthor;
 
-module.exports = (robot) => {
-  scheduler = createScheduler(robot, {
+/**
+ * This is the main entrypoint to your Probot app
```
Comment (ankita240796) [inline index.js:10]: your Probot app -> the Probot app.

File: lib/checkMergeConflicts.js [hunk 4]
```diff
@@ -5,6 +5,10 @@ function sleep(ms) {
   return new Promise(resolve => setTimeout(resolve, ms));
 }
 
+/**
+ * @param {import('probot').Context} context
+ * @param {import('probot').Octokit.PullsGetResponse} pullRequest
+ */
 module.exports.checkMergeConflictsInPullRequest = async function(context, pullRequest) {
```
Comment (ankita240796) [inline lib/checkMergeConflicts.js:12]: Line looks to long, split after bracket maybe.

**Comment by jameesjohn** [comment]
Screenshot (286)

@ankita240796 here's a screenshot of it in action.
It can also be viewed at https://github.com/jameesjohn/comment-on-pr/pull/15

File: index.js (hunk 3 above)
Comment (jameesjohn) [reply index.js]: Done

File: lib/checkMergeConflicts.js (hunk 4 above)
Comment (jameesjohn) [reply lib/checkMergeConflicts.js]: Done.

**Comment by jameesjohn** [comment]
Also, I noticed a bunch of console errors from https://github.com/oppia/oppiabot-stale. It should be updated as well.

**Comment by ankita240796** [comment]
> Also, I noticed a bunch of console errors from https://github.com/oppia/oppiabot-stale. It should be updated as well.

We should shift the stale check to probot scheduler itself instead of any separate upgrade. We should have all the oppiabot stuff in a single repo.

**Comment by jameesjohn** [comment]
Okay.

**Comment by jameesjohn** [comment]
PTAL @kevinlee12

**Comment by jameesjohn** [comment]
> It can also be viewed at jameesjohn/comment-on-pr#15

Here's a link to the [sheet used](https://docs.google.com/spreadsheets/d/1J4EP-SbxDBDAtnseTd7xn6WtPTyyO9a242bYvaUDOFs/edit#gid=0) for the testing.

File: package.json (hunk 1 above)
Comment (DubeySandeep) [reply package.json:20]: > It's not stated that's it's unsafe

I think `beta` keyword states it's unstable, I would also suggest you do what @ankita240796 has suggested and make sure to thoroughly test the feature we are using from this lib.

File: package.json [hunk 5]
```diff
@@ -6,6 +6,8 @@
   "license": "Apache 2.0",
   "repository": "https://github.com/oppiabot/oppiabot.git",
   "scripts": {
+    "dev": "nodemon",
```
Comment (DubeySandeep) [inline package.json:9]: is this a dependency for updated probot?

File: spec/oppiabotSpec.js [hunk 6]
```diff
@@ -9,9 +9,12 @@ const pullRequestpayload = require('../fixtures/pullRequestPayload.json');
 
 describe('Oppiabot\'s', () => {
   let robot;
+  /**
+   * @type {import('probot').Octokit} context
+   */
```
Comment (DubeySandeep) [inline spec/oppiabotSpec.js:14]: This is undefined here, right? So why are we adding type? Also, I see this variable is storing dict

File: index.js [hunk 7]
```diff
@@ -1,19 +1,22 @@
...
   (process.env.WHITELISTED_ACCOUNTS || '').toLowerCase().split(','));
-var pullRequestAuthor;
 
-module.exports = (robot) => {
-  scheduler = createScheduler(robot, {
+/**
+ * This is the main entrypoint to the Probot app
+ * @param {import('probot').Application} app
+ */
+module.exports = (app) => {
```
Comment (DubeySandeep) [inline index.js:13]: Any specific reason for changing robot to the app? (Just wanted to know the context behind this)

File: .gitignore [hunk 8]
```diff
@@ -3,3 +3,4 @@ npm-debug.log
 *.pem
 .env
 package-lock.json
+newrelic_agent.log
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply .gitignore]: Done

File: index.js (hunk 7 above)
Comment (jameesjohn) [reply index.js:13]: Not really.
That's what's used in the doc so I decided to do it.

File: package.json (hunk 5 above)
Comment (jameesjohn) [reply package.json:9]: No.
It's a dev dependency. Used to re-run the app when files change.

File: package.json (hunk 1 above)
Comment (jameesjohn) [reply package.json:20]: I've followed what @ankita240796 said by creating the issue.
Also, the feature to be used by this package is not in at the moment.
https://github.com/oppia/oppiabot/pull/54 is the first feature that uses probot scheduler.
While testing, I added little functionalities to check if it works, and it did.

Anyone else is welcome to test.
I also have a PR: jameesjohn/comment-on-pr#15 where the upgrade was being tested.

File: spec/oppiabotSpec.js (hunk 6 above)
Comment (jameesjohn) [reply spec/oppiabotSpec.js:14]: Adding the type helps in mocking the new functions in the probot upgrade as the editor will help with better IntelliSense.

Updated to show `github` instead of `context`

**Comment by jameesjohn** [comment]
Yes, @DubeySandeep. All the features have been manually tested.
I have a PR in the works about fixing the automated test, but that will be made after the upgrade is complete.
Thanks.

File: package.json (hunk 1 above)
Comment (jameesjohn) [reply package.json:20]: Also, probot v9 does not work with the older version of the scheduler.

Comment (ankita240796) [reply package.json:20]: I am fine with using beta version as long as we test everything. Did you run any check with this version and check if it works (since as of now we do not have any probot scheduler based checks).

**Comment by ankita240796** [comment]
Hey @DubeySandeep, can you please take a look, Thanks!

File: index.js (hunk 7 above)
Comment (DubeySandeep) [reply index.js:13]: In that case, I would suggest undoing this change. Considering the main repo app name is `oppia` so we should name it similarly (either make it `oppiabot` (or `robot`))

File: lib/checkMergeConflicts.js [hunk 9]
```diff
@@ -25,7 +30,7 @@ module.exports.checkMergeConflictsInPullRequest = async function(context, pullRe
   while (pullRequest.mergeable === null) {
     // Wait for 2 seconds before making another request.
     await sleep(2000);
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply lib/checkMergeConflicts.js:32]: Noted.
Thanks

File: index.js (hunk 7 above)
Comment (jameesjohn) [reply index.js]: `oppiabot` will suffice

File: lib/checkMergeConflicts.js [hunk 10]
```diff
@@ -68,18 +73,21 @@ module.exports.checkMergeConflictsInPullRequest = async function(context, pullRe
...
 
+/**
+ * @param {import('probot').Context} context
+ */
 module.exports.checkMergeConflictsInAllPullRequests = async function(context) {
   // pullRequests.getAll(..) fetches atmost 30 PRs in a batch, by default.
   // Since we've seen increasing number of PRs in the past, this value has
   // been increased to 60 PRs in a batch.
-  var pullRequestsPromiseObj = await context.github.pullRequests.getAll(
+  var pullRequestsPromiseObj = await context.github.pulls.list(
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply lib/checkMergeConflicts.js:83]: Okay, but I don't think that should be in this PR since it is not related to the upgrade.

File: lib/checkPullRequestLabels.js [hunk 11]
```diff
@@ -4,9 +4,15 @@ var matchChangelogLabelWithRegex = function(changelogLabel) {
...
   return regExpForChangelogLabel.test(changelogLabel);
-}
+};
 
+/**
+ * @param {import('probot').Context} context
+ */
 module.exports.checkAssignee = async function(context) {
+  /**
+   * @type {import('probot').Octokit.PullsGetResponse} context
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply lib/checkPullRequestLabels.js]: No
It's supposed to be pullRequest. Updated.

File: package.json [hunk 12]
```diff
@@ -6,6 +6,8 @@
...
+    "lint": "standard --fix",
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply package.json:10]: It is so that any auto-fixable issues found during the lint will be fixed automatically

[the remark this answers is not in the document]
Comment (jameesjohn) [reply package.json:10]: Just as a side note, I noticed that the rules set by standard and different and they go against themselves

File: package.json [hunk 13]
```diff
@@ -26,15 +28,22 @@
...
   "standard": {
     "env": [
       "jasmine"
     ]
+  },
+  "nodemonConfig": {
+    "exec": "npm start",
+    "watch": [
+      ".env",
+      "."
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply package.json:46]: By default, `nodemon` doesn't watch all files https://github.com/remy/nodemon#specifying-extension-watch-list.
This is done to add other files to be watched by `nodemon`.

File: spec/oppiabotSpec.js (hunk 6 above)
Comment (jameesjohn) [reply spec/oppiabotSpec.js:14]: No, that's the correct type.

File: package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (DubeySandeep) [reply package.json:10]: > I noticed that the rules set by standard and different and they go against themselves

Can you rephrase this?

> It is so that any auto-fixable issues found during the lint will be fixed automatically

Don't do this else CI (in future) will fix the issue in the local repo and it will assume that there isn't any (lint) issue in PR!

File: lib/checkMergeConflicts.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (DubeySandeep) [reply lib/checkMergeConflicts.js:83]: I see a method name in the comment which doesn't exist anymore so I wanted you to update the comment accordingly.

File: package.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (DubeySandeep) [reply package.json:46]: Can you explain what each of the item in the list covers? Also, is it auto-generated or you have updated it?

File: lib/checkMergeConflicts.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (jameesjohn) [reply lib/checkMergeConflicts.js:83]: @DubeySandeep , I've updated it already. The comment was for creating the constant.

File: package.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (jameesjohn) [reply package.json:46]: It is manually updated.
By default, nodemon doesn't watch .env and files so specifying it here tells nodemon to watch .env

File: package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (jameesjohn) [reply package.json]: > Don't do this else CI (in future) will fix the issue in the local repo and it will assume that there isn't any (lint) issue in PR!

Noted.

> Can you rephrase this?

I meant the linting rules set by `standard` contradicts those set by `eslint`

File: lib/checkMergeConflicts.js [hunk 14]
```diff
@@ -68,18 +73,21 @@ module.exports.checkMergeConflictsInPullRequest = async function(context, pullRe
   }
 };
 
+/**
+ * @param {import('probot').Context} context
+ */
 module.exports.checkMergeConflictsInAllPullRequests = async function(context) {
-  // pullRequests.getAll(..) fetches atmost 30 PRs in a batch, by default.
+  // pullRequests.list(..) fetches atmost 30 PRs in a batch, by default.
```
Comment (DubeySandeep) [inline lib/checkMergeConflicts.js:80]: This should be `pulls.list` (OR `context.github.pulls.list`)

Comment (DubeySandeep) [inline lib/checkMergeConflicts.js:80]: Also, can you please check whether this comment still persists with the updated probot?

File: package.json [hunk 15]
```diff
@@ -26,15 +28,22 @@
...
+    ]
   }
-}
+}
```
Comment (DubeySandeep) [inline package.json:49]: Add a newline (I would suggest you to config you editor to add a newline at the end of the file.)

File: package.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (DubeySandeep) [reply package.json:46]: So why do we need to watch the ignored file (.env) [Sorry if this question is silly but I've not worked in this project before.] Also "." in the watch list means to watch all the files and folder in the root dir?

**Comment by DubeySandeep** [comment]
@jameesjohn, There are 4 open comments above, other than that this PR LGTM! feel free to merge it once those comments are resolved!

File: lib/checkMergeConflicts.js (hunk 14 above)
Comment (jameesjohn) [reply lib/checkMergeConflicts.js]: Ahh
I missed it.
Thanks

Comment (jameesjohn) [reply lib/checkMergeConflicts.js]: Yes it does persist.

File: package.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (jameesjohn) [reply package.json]: Sometimes, you may need to make changes to the .env file like spreadsheet details  when testing the google spreadsheet api

[the remark this answers is not in the document]
Comment (jameesjohn) [reply package.json]: I've updated it to remove the "."

File: package.json (hunk 15 above)
Comment (jameesjohn) [reply package.json]: Done.
Thanks

**Comment by jameesjohn** [comment]
@DubeySandeep, I've responded to all comments. PTAL

**Comment by ankita240796** [comment]
Congrats on the upgrade @jameesjohn! Let us focus on fixing the tests now, in the meantime I will look into setting up CI for oppiabot repo.

**Comment by DubeySandeep** [comment]
Congrats and thanks for the upgrade, @jameesjohn!

Just wanted to highlight that we have to revisit the changes in package.json while adding tests in CI (like we may have to remove the watchers, linter etc.).

**Comment by ankita240796** [comment]
> Just wanted to highlight that we have to revisit the changes in package.json while adding tests in CI (like we may have to remove the watchers, linter etc.).

Noted, Thanks for the heads-up @DubeySandeep!

**Comment by jameesjohn** [comment]
Thanks @ankita240796 @DubeySandeep
